### PR TITLE
Feat/joi validation

### PR DESCRIPTION
**Pull request by danpatrickoneill**
# Description

Added data validation using Joi and a middleware function that will check the request body against our Joi schemas. Added schemas based on each current knex migration. Added imports and middleware call to user POST function, which serves as an example.

## Checklist

Remove any items which are not applicable.

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works AND the tests pass

### CODE REVIEW

File: middleware/dataValidation.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+// Takes in a schema, checks req body against it, and returns middleware that throws a 400 or moves on to the next
+module.exports = schema => {
+  return (req, res, next) => {
+    const { error } = Joi.validate(req.body, schema);
+    if (error) {
+      console.log(error);
+      return res.status(400).json({
+        error: `Error during ${req.method} at ${req.originalUrl}: ${
+          error.details[0].message
```
Comment (DaftBeowulf) [inline middleware/dataValidation.js]: Very descriptive message with the url as well, nice!

File: schemas/index.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    .required()
+});
+
+const groupUserSchema = Joi.object().keys({
+  user_id: Joi.number()
+    .integer()
+    .min(1)
+    .required(),
+  // Would like to add valid options below once we know what they are
+  user_type: Joi.string().required(),
```
Comment (DaftBeowulf) [inline schemas/index.js]: Definitely revisit this point once the user "types" get fleshed out, will need to be modified

File: schemas/index.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  allegianceSchema,
+  groupAllegianceSchema,
+  groupUserSchema,
+  postSchema,
+  postLikeSchema,
+  postTagSchema,
+  replySchema,
+  replyLikeSchema,
+  replyTagSchema
+};
```
Comment (DaftBeowulf) [inline schemas/index.js]: Well done getting all of these knocked out! Will make it very easy for the rest of the team to implement them

File: controllers/user.js [hunk 4]
```diff
@@ -4,13 +4,16 @@ const Users = require("../models/users");
...
+const { userSchema } = require("../schemas");
+
 router
   .route("/")
   .get(async (req, res) => {
     const users = await Users.find();
     res.status(200).json({ users });
   })
-  .post(async (req, res) => {
+  .post(validation(userSchema), async (req, res) => {
```
Comment (DaftBeowulf) [inline controllers/user.js]: This is a perfect example of easy to implement and use once the schemas have been set up -- it should be included for the PUT request below as well, though

File: middleware/dataValidation.js (hunk 1 above)
Comment (danpatrickoneill) [reply middleware/dataValidation.js]: Thank you! Mirrors the error message someone already had in the auth middleware, I believe.

File: schemas/index.js (hunk 2 above)
Comment (danpatrickoneill) [reply schemas/index.js]: Yep, and there is a good example of using valid() elsewhere in this file.

File: controllers/user.js (hunk 4 above)
Comment (danpatrickoneill) [reply controllers/user.js]: Somehow mis-targeted my merge on my machine, should be there now.

**Pull request merged**